developer/src/kmcmplib/src/CompilerInterfaces.cpp

```diff
@@ -133,105 +132,70 @@ EXTERN bool kmcmp_CompileKeyboardFile(char* pszInfile,
-
-  delete kmcmp::CodeConstants;
 
-  if (kmcmp::nErrors > 0)
-  {
-    remove(pszOutfile);
+  if (kmcmp::nErrors > 0) {
     return FALSE;
   }
 
   return result;
 }
 
-
-
 EXTERN bool kmcmp_CompileKeyboardFileToBuffer(char* pszInfile, void* pfkBuffer, bool ACompilerWarningsAsErrors, bool AWarnDeprecatedCode,
   kmcmp_CompilerMessageProc pMsgproc, void* AmsgprocContext, int Target)   // I4865   // I4866
 {
-  //printf("°°-> changed to CompileKeyboardFileToBuffer() of kmcmplib \n");
-  FILE* fp_in = NULL;
-  KMX_CHAR str[260];
-
-  kmcmp::FSaveDebug = TRUE;   // I3681
-  kmcmp::FCompilerWarningsAsErrors = ACompilerWarningsAsErrors;   // I4865
-  AWarnDeprecatedCode_GLOBAL_LIB = AWarnDeprecatedCode;
-  kmcmp::CompileTarget = Target;
-
   if (!pMsgproc || !pszInfile || !pfkBuffer) {
     SetError(CERR_BadCallParams);
   }
 
-  PKMX_STR p;
+  return CompileKeyboard(pszInfile, pfkBuffer, TRUE, ACompilerWarningsAsErrors, AWarnDeprecatedCode,
+    pMsgproc, AmsgprocContext, Target);
+}
 
-  if ((p = strrchr_slash(pszInfile)) != nullptr) {
-    strncpy(kmcmp::CompileDir, pszInfile, (int)(p - pszInfile + 1));  // I3481
-    kmcmp::CompileDir[(int)(p - pszInfile + 1)] = 0;
-  }
-  else {
-    kmcmp::CompileDir[0] = 0;
+EXTERN bool kmcmp_CompileKeyboardFile(char* pszInfile,
+  char* pszOutfile, bool ASaveDebug, bool ACompilerWarningsAsErrors,
+  bool AWarnDeprecatedCode, kmcmp_CompilerMessageProc pMsgproc, void* AmsgprocContext)   // I4865   // I4866
+{
+  if (!pMsgproc || !pszInfile || !pszOutfile) {
+    SetError(CERR_BadCallParams);
   }
 
-  msgproc = pMsgproc;
-  msgprocContext = AmsgprocContext;
-  kmcmp::currentLine = 0;
-  kmcmp::nErrors = 0;
-
-  fp_in = Open_File(pszInfile,"rb");
-
-  if (fp_in == NULL) {
-    SetError(CERR_InfileNotExist);
+  FILE_KEYBOARD fk;
+  if(!CompileKeyboard(pszInfile, &fk, ASaveDebug, ACompilerWarningsAsErrors, AWarnDeprecatedCode,
+      pMsgproc, AmsgprocContext, CKF_KEYMAN)) {
+    // any errors will have been reported directly by CompileKeyboard
+    return FALSE;
   }
 
-  // Transfer the file to a memory stream for processing UTF-8 or ANSI to UTF-16?
-  // What about really large files?  Transfer to a temp file...
-
-  if( !fread(str,1,3,fp_in))
-  {
-    fclose(fp_in);
-    SetError(CERR_CannotReadInfile);
+  FILE* fp_out = Open_File(pszOutfile, "wb");
+  if (fp_out == NULL) {
+    SetError(CERR_CannotCreateOutfile);
   }
 
-  fseek( fp_in,0,SEEK_SET);
-  if (str[0] == UTF8Sig[0] && str[1] == UTF8Sig[1] && str[2] == UTF8Sig[2])
-    fp_in = UTF16TempFromUTF8(fp_in, TRUE);
-  else if (str[0] == UTF16Sig[0] && str[1] == UTF16Sig[1])
-    fseek( fp_in,2,SEEK_SET);
-  else
-    fp_in = UTF16TempFromUTF8(fp_in, FALSE);
-
-  kmcmp::CodeConstants = new kmcmp::NamedCodeConstants;
+  KMX_DWORD msg;
+  KMX_BYTE* data = nullptr;
+  size_t dataSize = 0;
+  if ((msg = WriteCompiledKeyboard(&fk, &data, dataSize)) != CERR_None) {
+    AddCompileError(msg);
+  } else {
+    if(fwrite(data, 1, dataSize, fp_out) != dataSize) {
```

Comment:
What happens if we call `fwrite` if `fp_out` is NULL?

Reply:
`SetError` is a macro that emits `return FALSE;`, so if `fp_out == NULL` we'll never get here (line 176). I want to remove `SetError` because it is so misleading -- did the initial round of work in #8570.

Now is probably as good a time as any -- it's only used in a handful of places, all within CompilerInterfaces.cpp :grin: so expect a PR to arrive soon.